### ISSUE TITLE
 batch size, learning 

### ISSUE THREAD

**Issue by BigDataMLexplorer** [huggingface/peft#1948]
### System Info

LLM, fine tuning, neural networks, hyperparameters, batch size, learning rate, Huggingface

### Who can help?

@BenjaminBossan @sayakpaul @stevh 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### Reproduction

Hey, 
could I ask you for help? I am doing fine tuning of LLM model Llama3 8b (with LoRA) for text classification. I am using Trainer from Huggingface. I am looking for the optimal learning_rate and per_device_train_batch_size. 
I am using AdamW_torch optimizer and have over 80,000 training data. With per_device_train_batch_size = 2, one epoch takes about 12 hours. So it is hard for me to find optimal values because the training is long.
My best results so far are at 2 batch size, 1 epoch and learning rate 2e-4 --> 86.86% on validation data (about the same on test data). More epochs don't help, the model is starting to overfit whit this LR and batch size.

I am posting intermediate results here. Due to the long training time I do not send complete results (sometimes e.g. 1/3 of epochs etc..). Could someone give me an estimate based on these, what values of batch size and learning rate might help and what should be used for longer training? 

**1 EPOCH, LR 2e-4, PER_DEVICE_TRAIN_BATCH_SIZE = 2**
Epoch 0.5: {'eval_loss': 0.8900082111358643, 'eval_accuracy': 0.8452}
Epoch 1.0: {'eval_loss': 0.6152949929237366, 'eval_accuracy': 0.8686}

**2 EPOCHS, LR 2e-4, PER_DEVICE_TRAIN_BATCH_SIZE = 2**
Epoch 0.5: {'eval_loss': 0.9081119894981384, 'eval_accuracy': 0.8314}
Epoch 1.0: {'eval_loss': 0.7357500195503235, 'eval_accuracy': 0.8488}
Epoch 1.5: {'eval_loss': 0.7343335151672363, 'eval_accuracy': 0.8584}
Epoch 2.0: {'eval_loss': 0.5892598628997803, 'eval_accuracy': 0.8758} -> 87,1% on test data

**6 EPOCHS, LR 2e-4, PER_DEVICE_TRAIN_BATCH_SIZE = 2 (closed after 5 epochs)**
Step	Training Loss	Validation Loss Accuracy
41572	0.811400	0.872795	0.834000
83144	0.788500	0.774350	0.836600
124716	0.777800	0.737788	0.845200
166288	0.547900	0.732708	0.858400
207860	0.416800	0.746941	0.866000

**5 EPOCHS, LR 1e-5, PER_DEVICE_TRAIN_BATCH_SIZE = 2 
[ 84599/207860 26:28:07 < 38:33:56, 0.89 it/s, Epoch 2.03/5]**
Step	Training Loss	Validation Loss	Accuracy
20786	0.892400	0.824404	0.839000
41572	0.773900	0.792925	0.848000
62358	0.699800	0.842472	0.851800
83144	0.740500	0.713346	0.859400

**3 EPOCHS, LR 2e-5, PER_DEVICE_TRAIN_BATCH_SIZE = 16 
[13922/15591 12:29:10 < 1:29:49, 0.31 it/s, Epoch 2.68/3]**      
Step	Training Loss	Validation Los Accuracy
2598	1.575400	1.473688	0.540800
5196	1.209300	1.213021	0.619800
7794	1.135000	1.119776	0.642600            
10392	1.078100	1.078663	0.655800
12990	1.028700	1.062033	0.660600

**3 EPOCHS, LR 2e-5, PER_DEVICE_TRAIN_BATCH_SIZE = 32
 [1307/7797 2:19:22 < 11:33:09, 0.16 it/s, Epoch 0.50/3]**
Step	Training Loss	Validation Loss	Accuracy
1299	2.121100	1.649759	0.497800

**3 EPOCHS, **LR 16e-5**, PER_DEVICE_TRAIN_BATCH_SIZE = 16
[ 5222/15591 4:45:06 < 9:26:20, 0.31 it/s, Epoch 1.00/3]**
Step	Training Loss	Validation Loss	Accuracy
2598	**1.151000	1.053402**	**0.668400**           
5196	0.988100	1.030373	0.676200

**3 EPOCHS, **LR 16e-4**, PER_DEVICE_TRAIN_BATCH_SIZE = 16
[ 2608/15591 2:21:56 < 11:47:10, 0.31 it/s, Epoch 0.50/3]**
Step	Training Loss	Validation Loss	Accuracy
2598	**3.142300	2.734620**	  **0.642000**

**1 EPOCH, LR 2e-3, PER_DEVICE_TRAIN_BATCH_SIZE = 2
[31814/41572 8:05:48 < 2:29:00, 1.09 it/s, Epoch 0.77/1]**
Step	Training Loss	Validation Loss	Accuracy
20786	0.000000	nan	0.079200 -> 7% accuracy :) :) :) :), LR is too high

**1 EPOCH, LR 4e-4, PER_DEVICE_TRAIN_BATCH_SIZE = 2**
Epoch 0.5: {'eval_loss': 1.260171890258789, 'eval_accuracy': 0.6548,

**1 EPOCH, LR 1e-4, PER_DEVICE_TRAIN_BATCH_SIZE = 2
 [21434/41572 6:42:15 < 6:17:58, 0.89 it/s, Epoch 0.52/1]**
Step	Training Loss	Validation Loss	Accuracy
20786	0.864800	0.804458	0.851000

**1 EPOCH, LR 5e-5, PER_DEVICE_TRAIN_BATCH_SIZE = 2**
Epoch 0.5: 'eval_loss': 0.7843595743179321, 'eval_accuracy': 0.8532
Epoch 1.0: eval_loss': 0.6852059364318848, 'eval_accuracy': 0.8632

**My View:**
If I use batch size 2, learning rate 2e-4 and smaller, I can't see much progress. The results after running 1/1 are almost the same as after 5/5 (in 1/5 are worse than 1/1)
If I choose e.g. batch size 16, the results drop to values around 60%, then rise quickly, but stop again and rise slowly. With a larger batch size I get better intermediate results if I also increase the learning rate. I've been doing linear scaling --- for example, Batch from 2 to 16 --> original LR * 8.

Isn't this weird? -> If I choose this: 
3 EPOCHS, LR **16e-5**, PER_DEVICE_TRAIN_BATCH_SIZE = 16, I have 66.8% val accuracy with 1.151000 training loss and 1.053402 val loss at mid epoch, but if I put LR **16e-4**, I have very similar accuracy with (64.2%) but 3.142300 training loss and 2.734620 val loss. What does this mean please??? 

1) Does it mean that this model with more train and val loss would have more potential to improve?
2) Should I scale the laearning rate linearly with increasing batch size in AdamW-torch optimizer? 

I considered using tools like Optuna or Rey Tune to find the best hyperparameters. But would it be worth it at all if the training takes this long and isn't it better to try it myself? 

**My main question is what would you choose as optimal parameters for longer training as I am not very experienced with tuning hyperparameters.**

I would be grateful to anyone who can help me. Thank you.

### Expected behavior

.

**Comment by BigDataMLexplorer** [comment]
Sorry, this is maybe more related to transformers or general problem.

**Comment by BenjaminBossan** [comment]
It is very hard to give general recommendations based on what you report. I think your main intuition to increase the learning rate when the batch size is increased is correct and is usually applied when training neural nets (i.e. this is independent of whether PEFT is used or not).

With PEFT, it is also common to choose a higher learning rate than when fully fine-tuning. However, it looks like you're already doing that, as ~2e-4 looks like the right ballpark.

If you post your model and your `LoraConfig`, I might also be able to check if I see some hyper-parameters that you could try. Doing a full hyper-parameter sweep is, however, very expensive as you noted. You could try that with a smaller dataset, but it's hard to say if the results would generalize to the full dataset.

> Isn't this weird? -> If I choose this:
> 3 EPOCHS, LR **16e-5**, PER_DEVICE_TRAIN_BATCH_SIZE = 16, I have 66.8% val accuracy with 1.151000 training loss and 1.053402 val loss at mid epoch, but if I put LR **16e-4**, I have very similar accuracy with (64.2%) but 3.142300 training loss and 2.734620 val loss. What does this mean please???

First of all, what are scores that you would expect? Do you have reason to believe that much better validation scores could be obtained? Sometimes, it helps to take a step back and investigate what a good score would even be.

For your observation that accuracy can be quite good even when the loss is relatively high, this depends on the data. A model can have a high accuracy even if the predicted probability for the correct class is very small, as long as it's the highest probability. I would assume that the models with better loss should have better calibrated probabilities (as long as they're not overfitting). So if you check some scores that are taking probabilities into account, you will probably find that they are better when loss is smaller.

> Sorry, this is maybe more related to transformers or general problem.

Yes, in general the issue tracker is more for bug reports, feature requests etc. For discussions, check:

- Forum: https://discuss.huggingface.co/
- Discord: http://hf.co/join/discord
- GitHub discussions: https://github.com/huggingface/peft/discussions (but not a lot of activity here)

**Comment by BigDataMLexplorer** [comment]
Thank you very much for your reponse.

This is how I load the model: model = `AutoModelForSequenceClassification.from_pretrained("path", num_labels=11 ,device_map="auto")`
And this is my LoRA configuration:
```
lora_config = LoraConfig(
    r = 16, 
    lora_alpha = 8, 
    target_modules = ['q_proj', 'k_proj', 'v_proj', 'o_proj'],
    lora_dropout = 0.05, 
    bias = 'none', 
    task_type = 'SEQ_CLS'
```
)
I kept the LoRA configuration unchanged during each training session. So all the results I sent are with this configuration
I also used these other hyperparameters:
weight_decay=0.01
optim=Adamw_torch
fp16=True
 warmup_ratio=0.1
(gradient_checkpointing=True only for larger batches)          

#First of all, what are scores that you would expect? Do you have reason to believe that much better validation scores could be obtained?   
I've read that usually smaller batches have a better effect on model generalization on unseen data. However, I often came across the fact in peer reviewed studies that batch 32 was often the best choice for resulting accuracy than smaller batches. So a couple of times in the results above I have given intermediate results for batch 16 or 32. To do this, I linearly scaled LR. I would like to achieve accuracies above 90%, and I need to estimate which is the best choice based on my parameters and the results I've posted. For example, it can be seen that for higher batch sizes the accuracy starts above 60% and then improves very slowly. In one experiment , as I mention, I have the same accurcy but higher train and val loss with the same higher batch size and different LR. I was wondering if this is a better option that the training would improve quickly. Since the training is long and expensive I can't let the failed training run for long. 
My question is if I can't somehow make the best guess on the results above because I don't understand it and don't know who to contact. Would you please have any ideas or contacts?

Thank you very much

**Comment by BenjaminBossan** [comment]
I can give you a couple of hints but at the end of the day, the only way I could really help is if you gave me access to the data and a couple of days to study the problem ;-)

1. Your task seems to be sequence classification. After the `get_peft_model` call, ensure that the model's classification head is wrapped in a `ModulesToSaveWrapper` (so that it is fully fine-tuned).
2. `target_modules = ['q_proj', 'k_proj', 'v_proj', 'o_proj'],`: I don't know what base model you use, but this is a helpful hyper-parameter to tune (e.g. are there other linear layers that could be targeted).
3. `r = 16`: Also worth trying higher values.
4. `lora_alpha = 8`: Try setting `lora_alpha` to twice the value of `r`. You could also try passing `use_rslora=True`.
5. `fp16=True`: Generally, bf16 is better than fp16, so use that if your GPU supports it.
6. "I would like to achieve accuracies above 90%": Do you know if it's possible? Did you run a simple model as a benchmark (e.g. TFIDF + logistic regression)?
7. "Since the training is long and expensive I can't let the failed training run for long.": How frequently do you report results? If only very slowly on an epoch level, consider reporting results every few batches instead so that you can get quick feedback and abort runs that look bad right at the start.
8. Generally, getting a quicker iteration cycle should be high priority for you. For instance, you could try working with a small, representative subsample of your dataset first to figure out best settings and only then run on the whole dataset. But there is no guarantee that this will generalize.
9. "My question is if I can't somehow make the best guess on the results above": I don't think there is really a way here.

**Comment by BigDataMLexplorer** [comment]
@BenjaminBossan 
Thank you for your reply!
1) Yes, it is. I classify the company's customer complaints into 11 categories.
2) The base model is Llama3 8b
4) Ok, I will try lora_alpha to 32, thank you for your tip
5) Yes I have gpu that supports only fp16
6) Our TFIDF solution has 86% accuracy and we are thinking about using Large Langue Model
7) I validate the training in the middle of each epoch. I know that at this point it is already showing good results. I don't know 
    how to choose the exact epoch.
8) It's hard to say. I've got 83,000 records in dataset. It's hard to know what a representative sample is.
Unfortunately, this is sensitive company data and I cannot send it.

**Comment by BenjaminBossan** [comment]
> 2\. The base model is Llama3 8b

In this case, you could try to also target the gate projections of the MLP part. Just pass `target_modules="all-linear"` and you should be good.

> 4\. Yes I have gpu that supports only fp16

Okay. If you have enough memory, you could try fp32 instead.

> 5\. Our TFIDF solution has 86% accuracy and we are thinking about using Large Langue Model

I see, so roughly the same accuracy. What you could check if the TFIDF model and the LLM make the same mistakes, i.e. the same incorrect classification on the test set. If yes, maybe these samples are just too hard or incorrectly labeled. If they do different mistakes, it means there is potential for a better model. 

> 6\. I don't know how to choose the exact epoch.

Depends on what you use for training. If you use the transformers `Trainer`, you can change the `logging_strategy` and `logging_steps` of the [`TrainingArguments`](https://huggingface.co/docs/transformers/v4.43.2/en/main_classes/trainer#transformers.TrainingArguments) to change how often results are reported.

> 7\. It's hard to know what a representative sample is.

You could try a stratified random split. Since you already have some finished training runs, you could check if those splits are good predictors of the final results you would get on the whole dataset.

I don't think there is much more I can help you with, good luck.

**Comment by BigDataMLexplorer** [comment]
Thanks for your tips and help. I will consider the advice given. I will also put this post in disccusion where it fits better.
